subsys/bluetooth/mesh/time_srv.c

```diff
@@ -0,0 +1,491 @@
+	.settings_set = bt_mesh_time_srv_settings_set,
+#endif
+};
+
+int bt_mesh_time_srv_time_status_send(struct bt_mesh_time_srv *srv,
+				      struct bt_mesh_msg_ctx *ctx,
+				      struct bt_mesh_time_status *status)
+{
+	BT_MESH_MODEL_BUF_DEFINE(msg, BT_MESH_TIME_OP_TIME_STATUS,
+				 BT_MESH_TIME_MSG_LEN_TIME_STATUS);
+	update_state(&srv->data);
+	time_encode_time_params(&msg, status, BT_MESH_TIME_OP_TIME_STATUS);
+
+	srv->model->pub->ttl = 0;
+	return model_send(srv->model, ctx, &msg);
+}
+
+int bt_mesh_time_srv_get_tai_status(struct bt_mesh_time_srv *srv, u64_t uptime,
+				    struct bt_mesh_time_status *time)
+{
+	if (srv->data.tai_sec == TIME_UNKNOWN) {
+		return -ENOTSUP;
+	}
+
+	update_state(&srv->data);
+	s64_t delta_ms = uptime - srv->data.sync_uptime;
+	s64_t tai_tot_ms = (srv->data.tai_sec * MSEC_PER_SEC) +
+			   srv->data.subsec_ms + delta_ms;
+	if (tai_tot_ms < 0) {
+		return -EAGAIN;
+	}
+
+	time->tai_sec = tai_tot_ms / MSEC_PER_SEC;
+	time->subsec = ms_to_subsec_conv(tai_tot_ms % MSEC_PER_SEC);
+	time->uncertainty =
+		srv->data.uncertainty_base + srv->data.uncertainty_drift;
+	time->is_authority = srv->data.is_authority;
+
+	if ((srv->data.tai_of_zone_change == TIME_UNKNOWN) ||
+	    ((time->tai_sec < srv->data.tai_of_zone_change) &&
+	     (srv->data.tai_of_zone_change < srv->data.tai_sec))) {
+		return -EAGAIN;
+	} else if (time->tai_sec >= srv->data.tai_of_zone_change) {
+		time->time_zone_offset = srv->data.time_zone_offset_new;
+	} else {
+		time->time_zone_offset = srv->data.time_zone_offset_current;
+	}
+
+	if ((srv->data.tai_of_delta_change == TIME_UNKNOWN) ||
+	    ((time->tai_sec < srv->data.tai_of_delta_change) &&
+	     (srv->data.tai_of_delta_change < srv->data.tai_sec))) {
+		return -EAGAIN;
+	} else if (time->tai_sec >= srv->data.tai_of_delta_change) {
+		time->tai_utc_delta = srv->data.tai_utc_delta_new;
+	} else {
+		time->tai_utc_delta = srv->data.tai_utc_delta_current;
+	}
+	return 0;
+}
+
+int bt_mesh_time_srv_get_uptime(struct bt_mesh_time_srv *srv, u64_t tai_sec,
+				u8_t subsec, u8_t time_zone_offset,
+				u64_t *uptime)
+{
+	update_state(&srv->data);
+	u64_t current_tai_ms =
+		(srv->data.tai_sec * MSEC_PER_SEC) + srv->data.subsec_ms;
+	s64_t target_tai_ms = (tai_sec * MSEC_PER_SEC) +
+			      subsec_to_ms_conv(subsec) +
+			      ((time_zone_offset - (u64_t)ZONE_OFFSET_ZERO) *
+			       (u64_t)TIME_ZONE_OFFSET_UNIT_STEP);
+
+	if (target_tai_ms < 0) {
+		return -ECANCELED;
+	}
+	s64_t delta = target_tai_ms - current_tai_ms;
+
+	if ((delta < 0) && (srv->data.sync_uptime < abs(delta))) {
+		return -ECANCELED;
+	}
+	*uptime = srv->data.sync_uptime + delta;
+	return 0;
+}
+
+void bt_mesh_time_srv_time_set_manual(struct bt_mesh_time_srv *srv,
+				      struct bt_mesh_time_srv_data *data)
+{
+	memcpy(&srv->data, data, sizeof(struct bt_mesh_time_srv_data));
+	srv->data.sync_uptime = k_uptime_get();
+	srv->data.tai_sec_last_update = data->tai_sec;
+	store_state(srv);
+}
```

Comment:
Needs a role check or setting?

Reply:
Not quite sure what you mean. Do you mean like checking if it is a Mesh Time Authority or Mesh Time Relay? As it is now the user will overwrite the entire time context, and must ensure that the input structure is properly initialized. That includes the role of the server. Can you elaborate?